### ISSUE TITLE
Get Content-ID Response Header from A Batch Response In A Changeset

### ISSUE THREAD

**Issue by SpaceCondor** [SAP/cloud-sdk-java#344]
**Is your feature request related to a problem? Please describe.**
It doesn't seem to be possible to get Content-ID (or other arbitrary headers) of a batch response. For example, if I create a changeset like this:

```
BatchRequestBuilder batchRequestBuilder = new BatchRequestBuilder("/v1/");
        batchRequestBuilder.addChangeset(createDocument1, createDocument2, createDocument3));

BatchResponse batchResponse = batchRequestBuilder.execute(destination);
```
I can see the data in the request/response log like the following:

```
[Thu Feb 29 22:06:11 2024] [10.0.0.2] [pid=4152] [Request] "POST /v1/$batch HTTP/1.1"
OData-Version:4.0
Content-Type:multipart/mixed;boundary=batch_17d7a8c3-d3c7-42d4-b6a6-06ef1d975008
Content-Length:2173
Host:xxxxxxxxx.com:50000
Connection:Keep-Alive
User-Agent:Apache-HttpClient/4.5.14 (Java/21.0.2)
Cookie:**************************************************************************
Accept-Encoding:gzip,deflate
X-Forwarded-Proto:https

--batch_17d7a8c3-d3c7-42d4-b6a6-06ef1d975008
Content-Type: multipart/mixed;boundary=changeset_d3c4fa68-582a-46fd-8691-d24272625cda

--changeset_d3c4fa68-582a-46fd-8691-d24272625cda
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 1

POST Orders HTTP/1.1
Accept: application/json
Content-Type: application/json

{"@odata.type":"#SAPB1.Document"..........}

--changeset_d3c4fa68-582a-46fd-8691-d24272625cda
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 2

POST Orders HTTP/1.1
Accept: application/json
Content-Type: application/json

{"@odata.type":"#SAPB1.Document"..........}

--changeset_d3c4fa68-582a-46fd-8691-d24272625cda
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 3

POST Orders HTTP/1.1
Accept: application/json
Content-Type: application/json

{"@odata.type":"#SAPB1.Document"..........}

--changeset_d3c4fa68-582a-46fd-8691-d24272625cda--

--batch_17d7a8c3-d3c7-42d4-b6a6-06ef1d975008--


[Thu Feb 29 22:06:11 2024] [10.0.0.2] [pid=4152] [Response] "POST /v1/$batch HTTP/1.1"
Status Code:[200 OK]
Date:Thu, 29 Feb 2024 22:06:11 GMT
Server:Apache
OData-Version:4.0
Content-Length:569
Content-Type:multipart/mixed;boundary=batchresponse_hKXEHg8m-qSXb-TCW6-dl3M-4r2H6qxgAbaG

--batchresponse_hKXEHg8m-qSXb-TCW6-dl3M-4r2H6qxgAbaG
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 3

HTTP/1.1 400 Bad Request
Content-Type: application/json;charset=utf-8
Content-Length: 258
OData-Version: 4.0

{
   "error" : {
      "code" : "-5002",
      "details" : [
         {
            "code" : "",
            "message" : ""
         }
      ],
      "message" : "Error."
   }
}

--batchresponse_hKXEHg8m-qSXb-TCW6-dl3M-4r2H6qxgAbaG--
```

It does not seem to be possible (currently) to grab the Content-ID of the batch response message.

For example, I can see that the Content-ID 3 is in the Batch Response. This is helpful for knowing _which_ operation in a changeset caused the failure. However, I don't see a way to grab this.

**Comment by CharlesDuboisSAP** [comment]
Hello @SpaceCondor,

### Good news
Failed OData v4 Batch requests will return the specific failed request from the exception:
```java
try {
    batchResponse.getResult(create2);
}
catch( ODataResponseException e ) {
    e.getRequest();
}
```
It will be available at the next release `5.6.0`.

<hr/>

### Bad news

Our investigation found that your OData server response does not follow the [OData protocol](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_MultipartBatchResponse):
> The body of a multipart response to a multipart batch request MUST structurally match one-to-one with the multipart batch request body

And

> When a request within a change set fails, the change set response [...] is returned that applies to all requests in the change set

Meaning that your OData response should have a `changeset` boundary:
```diff
   --batchresponse_hKXEHg8m-qSXb-TCW6-dl3M-4r2H6qxgAbaG
++ Content-Type: multipart/mixed; boundary=changesetresponse_4459f831-1259-4f7c-89a0-95162d172489

++ --changesetresponse_4459f831-1259-4f7c-89a0-95162d172489
   Content-Type: application/http
   Content-Transfer-Encoding: binary
   Content-ID: 3

   HTTP/1.1 400 Bad Request
   Content-Type: application/json;charset=utf-8

   {"error"}
++ --changesetresponse_4459f831-1259-4f7c-89a0-95162d172489--
   --batchresponse_hKXEHg8m-qSXb-TCW6-dl3M-4r2H6qxgAbaG--
```
You should talk to the owner of your OData server to fix the response.

Best regards,
Charles

**Comment by SpaceCondor** [comment]
@CharlesDuboisSAP That is great news!

This is actually an SAP Server, it is the Service Layer for SAP Business One. I'll open an incident with them to notify them.

**Comment by CharlesDuboisSAP** [comment]
@SpaceCondor

Very good news!
Upon further investigation we are taking into account the behavior of SAP systems and it should be fixed for you in `5.6.0` :tada:

**Comment by newtork** [comment]
`5.6.0` is released and available for consumption now.

**Comment by andybai** [comment]
@CharlesDuboisSAP 

Could you please further check the OData spec as below:

_11.7.7.5 Multipart Batch Response
A multipart response to a batch request MUST contain a Content-Type header with value multipart/mixed.
The body of a multipart response to a multipart batch request MUST structurally match one-to-one with the multipart batch request body, such that the same multipart message structure defined for requests is used for responses. There are three exceptions to this rule:
·         **When a request within a change set fails, the change set response is not represented using the multipart/mixed media type. Instead, a single response, using the application/http media type, is returned that applies to all requests in the change set and MUST be a valid OData error response.**
·         When an error occurs processing a request and the continue-on-error preference is not specified, or specified with an explicit value of false, processing of the batch is terminated and the error response is the last part of the multipart response._
·       

As I understand it, the multipart/mixed media type should not be present in case error happens in the change set. So I suspect this is not an issue in the OData service.
Thanks.

**Comment by MatKuhr** [comment]
Hi @andybai, reading through the OData specification again I think you are right. While the overall response is still a multipart/mixed, the changeset itself isn't. So agreed, not an issue with the service. Unfortunately, the specification doesn't contain an example, but I found one from Microsoft [here](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/execute-batch-operations-using-web-api#handling-errors).

But since we adapted the Cloud SDK with the above change it now works with both formats. The implementation also serves for OData 2.0 responses, where no strict specification exists, so we'll probably still keep the old behavior around in parallel.
